proto/public/api.proto

```diff
@@ -54,6 +54,50 @@ message TapRequest {
   string path = 11;
 }
 
+// A tap request over kubernetes resources.
+message K8sTapRequest {
+  // Describes the kubernetes pods that should be tapped.
+  ResourceSelection target = 1;
+
+  // Selects over events to be reported.
+  Match match = 2;
+
+  // Limits the number of events to be inspected.
+  float maxRps = 3;
+
+  message Match {
+    oneof match {
+      // If empty, matches all messages.
+      Seq all = 1;
+
+      // If empty, matches no messages.
+      Seq any = 2;
+
+      // Inverts the inner match.
+      Match not = 3;
+
+      // Matches events being sent to any of the selected destinations.
+      ResourceSelection destinations = 4;
+
+      // Matches HTTP requests by their metadata.
+      Http http = 5;
+    }
+
+    message Seq {
+      repeated Match matches = 1;
```

Comment:
The recursive aspect of the match struct feels a bit convoluted to me, but it's entirely possible that I don't understand all of the use cases.

```
message Match {
  oneof match {
    Seq all = 1;
    ...
  }
  message Seq {
    repeated Match matches = 1;
  }
}
```

Translating this to words... the `K8sTapRequest` message includes a match selector. The match selector can contain a series of other match selectors, which can contain a series of other match selectors. But eventually all match selectors must be either a resource selector or an http selector. Is that basically correct? I'm not totally convinced that we need this level of flexibility, and it feels like might be pretty hard to reason about when we wire up the API.

Reply:
I exposed this API here because all of this functionality is already supported by the proxy.  My hope is that this doesn't actually add complexity to the controller, as it should really just be translating between this API and the proxy tap api, which already exposes these same primitives.

It's intended to allow this sort of thing:

_Tap either `POST` and `PUT` messages to _MyDst_)
```
Match::all(Match::Seq(
  Match::destinations(MyDst),
  Match::any(Match::Seq(
    Match::Http::method("POST"),
    Match::Http::method("PUT")
  ))
))
```

This would (roughly) get translated into a proxy match like:

```
Match::all(Match::Seq(
  Match::label("service", "mysvc"),
  Match::any(Match::Seq(
    Match::Http::method("POST"),
    Match::Http::method("PUT")
  ))
))
```

Reply:
Cool, works for me -- thanks for the explanation.